java/semantickernel-api/src/main/java/com/microsoft/semantickernel/FunctionInvocation.java

```diff
@@ -0,0 +1,191 @@
+            .withPromptExecutionSettings(promptExecutionSettings)
+            .withToolCallBehavior(toolCallBehavior);
+    }
+
+    public FunctionInvocation<T> withKernelHooks(
+        @Nullable
+        KernelHooks hooks) {
+        this.hooks = unmodifiableClone(hooks);
+        return this;
+    }
+
+    public FunctionInvocation<T> withPromptExecutionSettings(
+        @Nullable
+        PromptExecutionSettings promptExecutionSettings) {
+        this.promptExecutionSettings = promptExecutionSettings;
+        return this;
+    }
+
+    public FunctionInvocation<T> withToolCallBehavior(@Nullable ToolCallBehavior toolCallBehavior) {
+        this.toolCallBehavior = unmodifiableClone(toolCallBehavior);
+        return this;
+    }
+
+    @Nullable
+    private static UnmodifiableToolCallBehavior unmodifiableClone(
+        @Nullable
+        ToolCallBehavior toolCallBehavior) {
+        if (toolCallBehavior instanceof UnmodifiableToolCallBehavior) {
+            return (UnmodifiableToolCallBehavior) toolCallBehavior;
+        } else if (toolCallBehavior != null) {
+            return toolCallBehavior.unmodifiableClone();
+        } else {
+            return null;
+        }
+    }
+
+    @Nullable
+    private static UnmodifiableKernelHooks unmodifiableClone(
+        @Nullable KernelHooks kernelHooks) {
+        if (kernelHooks instanceof UnmodifiableKernelHooks) {
+            return (UnmodifiableKernelHooks) kernelHooks;
+        } else if (kernelHooks != null) {
+            return kernelHooks.unmodifiableClone();
+        } else {
+            return null;
+        }
+    }
+
+    @Override
+    public void subscribe(CoreSubscriber<? super FunctionResult<T>> coreSubscriber) {
+        performSubscribe(
+            coreSubscriber,
+            kernel,
+            function,
+            arguments,
+            resultType,
+            new InvocationContext(
+                hooks,
+                promptExecutionSettings,
+                toolCallBehavior
+            ));
+    }
+
+    // Extracted to static to ensure mutable state is not used
+    private static <T> void performSubscribe(
+        CoreSubscriber<? super FunctionResult<T>> coreSubscriber,
+        Kernel kernel,
+        KernelFunction<?> function,
+        @Nullable
+        KernelFunctionArguments arguments,
+        @Nullable
+        ContextVariableType<T> variableType,
+        @Nullable
+        InvocationContext context
+    ) {
+        if (variableType == null) {
+            LOGGER.debug(
+                "No variable type explicitly specified by calling 'withResultType' for function invocation: "
+                    + function.getSkillName() + "." + function.getName() + "."
+                    + " This may cause a runtime error (probably a ClassCastException) if the result type is not compatible with the expected type.");
+        }
+
+        function.invokeAsync(
+                kernel,
+                arguments,
+                null,
```

Comment:
Why are you not passing variableType here? Won't KernelFunction#invokeAsync still need to handle this argument in the manner of convertToType?

Reply:
invokeAsync no longer performs any type conversion to the required return type, by passing null here we effectively allow it to return whatever it naturally would

Reply:
also note that in the absence of an explicit return type the functions own metadata return type is used